llama-index-core/llama_index/core/workflow/checkpointer.py

```diff
@@ -0,0 +1,180 @@
+import asyncio
+import uuid
+import functools
+from llama_index.core.bridge.pydantic import (
+    BaseModel,
+    Field,
+    ConfigDict,
+)
+from typing import (
+    Optional,
+    Dict,
+    Any,
+    List,
+    Protocol,
+    TYPE_CHECKING,
+    Type,
+    Awaitable,
+)
+from llama_index.core.workflow.context import Context
+from llama_index.core.workflow.context_serializers import BaseSerializer, JsonSerializer
+from llama_index.core.workflow.handler import WorkflowHandler
+from .events import Event
+from .errors import *
+
+if TYPE_CHECKING:  # pragma: no cover
+    from .workflow import Workflow
+
+
+class CheckpointCallback(Protocol):
+    def __call__(
+        self,
+        last_completed_step: Optional[str],
+        input_ev: Optional[Event],
+        output_ev: Event,
+        ctx: Context,
+    ) -> Awaitable[None]:
```

Comment:
If it returns an awaitable, shouldn't this be an async function? Or what should be returned here?

Reply:
Ohh this is just to define a [Callback Protocol](https://mypy.readthedocs.io/en/stable/protocols.html#callback-protocols) but the async variant, for which the return type should be an `Awaitable` -- after a google search I found how out how to do this (see [here](https://github.com/python/mypy/issues/9328)))